core/src/main/java/org/apache/calcite/rel/RelDistributions.java

```diff
@@ -135,10 +141,14 @@ public RelDistribution apply(Mappings.TargetMapping mapping) {
       if (keys.isEmpty()) {
         return this;
       }
-      return getTraitDef().canonize(
-          new RelDistributionImpl(type,
-              ImmutableIntList.copyOf(
-                  Mappings.apply2((Mapping) mapping, keys))));
+      for (int key : keys) {
+        if (mapping.getTargetOpt(key) == -1) {
+          return RANDOM_DISTRIBUTED; // Some distribution keys are not mapped => random.
```

Comment:
I would recommend `ANY`, since it is the default.

Reply:
Fixed